### PR TITLE
Change retry support implementation

### PR DESCRIPTION
**Pull request by ssbarnea**
- remove dependency on tenacity
- retry support will be disabled by default
- adds 'retry' argument to Runtime constructor to opt-in for retry logic
- use retry only on galaxy install commands by default

Fixes: #26

### CODE REVIEW

File: src/ansible_compat/runtime.py [hunk 1]
```diff
@@ -113,16 +119,29 @@ def clean(self) -> None:
...
-            check=False,
-            stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE,
-        )
+    def exec(
+        self, args: Union[str, List[str]], retry: bool = False
+    ) -> CompletedProcess:
+        """Execute a command inside an Ansible environment.
+
+        param: retry: Retry network operations on failures.
```
Comment (webknjaz) [inline src/ansible_compat/runtime.py:127]:
```suggestion
        :param retry: Retry network operations on failures.
```

File: src/ansible_compat/runtime.py [hunk 2]
```diff
@@ -113,16 +119,29 @@ def clean(self) -> None:
...
-            stderr=subprocess.PIPE,
-        )
+    def exec(
+        self, args: Union[str, List[str]], retry: bool = False
+    ) -> CompletedProcess:
+        """Execute a command inside an Ansible environment.
+
+        :param retry: Retry network operations on failures.
+        """
+        for _ in range(self.max_retries if retry else 1):
```
Comment (webknjaz) [inline src/ansible_compat/runtime.py:129]: Why not
```suggestion
        for _ in range(self.max_retries if self.retry else 1):
```

File: src/ansible_compat/runtime.py [hunk 3]
```diff
@@ -165,12 +184,14 @@ def install_collection(
         cmd.append(f"{collection}")
 
         _logger.info("Running %s", " ".join(cmd))
-        run = self.exec(cmd)
+        run = self.exec(
+            cmd,
+            retry=self.retry,
```
Comment (webknjaz) [inline src/ansible_compat/runtime.py:189]: It feels weird having to pass this flag to both the class initializer and the individual method.

File: src/ansible_compat/runtime.py (hunk 2 above)
Comment (ssbarnea) [reply src/ansible_compat/runtime.py:129]: You indeed spot an design issue but that is not really the correct fix. The default retry value used by exec should be self.retry but we should still keep it. Users may want to alter the retry behavior for specific commands, even inside the library itself we do this. Mainly galaxy commands should use retry but other operations should not.

Let's check the use cases to get it right from start.

File: src/ansible_compat/runtime.py [hunk 4]
```diff
@@ -112,16 +116,29 @@ def clean(self) -> None:
...
-            check=False,
-            stdout=subprocess.PIPE,
-            stderr=subprocess.PIPE,
-        )
+    def exec(
+        self, args: Union[str, List[str]], retry: bool = False
+    ) -> CompletedProcess:
+        """Execute a command inside an Ansible environment.
+
+        :param retry: Retry network operations on failures.
```
Comment (webknjaz) [inline src/ansible_compat/runtime.py:124]: nit: why isn't `args` documented?

Comment (ssbarnea) [reply src/ansible_compat/runtime.py:124]: I was not able to find a nice way to document the args/kwargs on run, as we mainly pass them to `subprocess.run` but I agree that we should expose those that matter.

**Pull request merged**